p2p/src/swarm/peerdb.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): A. Altonen
+
+//! Peer database
+//!
+//! The peer database recognizes three peer types:
+//! - active peers
+//! - idle peers
+//! - reserved peers (not implemented)
+//!
+//! Active peers are those peers that the [`crate::swarm::PeerManager`] has an active connection with
+//! whereas idle peers are the peers that are known to `PeerDb` but are not part of our swarm.
+//! Idle peers are discovered through various peer discovery mechanisms and they are used by
+//! [`crate::swarm::PeerManager::heartbeat()`] to establish new outbound connections if the actual
+//! number of active connectios is less than the desired number of connections.
+//!
+//! TODO: reserved peers
+
+use crate::net::{types, NetworkingService};
+use logging::log;
+use std::collections::{HashMap, HashSet};
+
+// TODO: store active address
+// TODO: store other discovered addresses
+#[derive(Debug)]
+pub struct PeerContext<T: NetworkingService> {
+    pub _info: types::PeerInfo<T>,
+}
+
+/// Peer address information
+enum PeerAddrInfo<T: NetworkingService> {
+    Raw {
+        /// Hashset of IPv4 addresses
+        ip4: HashSet<T::Address>,
+
+        /// Hashset of IPv6 addresses
+        ip6: HashSet<T::Address>,
+    },
+}
+
+// TODO: improve how peers are stored, order by reputation?
+pub struct PeerDb<T: NetworkingService> {
```

Comment:
It's also necessary to store those on disk. But that's probably something can be done later. Bitcoin uses a custom format for storing peers in the data directory. If you want to go through with that, please make sure it's mockable.

Reply:
How about SCALE-encoded binary blobs, wouldn't that be lovely. I created an issue, #241 and I'll resolve this for now.